service/common/src/main/java/org/apache/polaris/service/events/listeners/FileBufferPolarisPersistenceEventListener.java

```diff
@@ -0,0 +1,189 @@
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import org.apache.polaris.core.config.FeatureConfiguration;
+import org.apache.polaris.core.config.PolarisConfigurationStore;
+import org.apache.polaris.core.context.CallContext;
+import org.apache.polaris.core.context.RealmContext;
+import org.apache.polaris.core.entity.PolarisEvent;
+import org.apache.polaris.core.persistence.BasePersistence;
+import org.apache.polaris.core.persistence.MetaStoreManagerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Event listener that buffers in memory and then dumps to persistence. */
+@ApplicationScoped
+@Identifier("persistence-file-buffer")
+public class FileBufferPolarisPersistenceEventListener extends PolarisPersistenceEventListener {
+  private static final Logger LOGGER =
+      LoggerFactory.getLogger(FileBufferPolarisPersistenceEventListener.class);
+  MetaStoreManagerFactory metaStoreManagerFactory;
+  PolarisConfigurationStore polarisConfigurationStore;
+  Clock clock;
+
+  // Key: str - Realm
+  // Value:
+  //    Key: int - shard number
+  //    Value: BufferShard - an object representing the directory and file where events are
+  // persisted on the filesystem
+  private final HashMap<String, HashMap<Integer, BufferShard>> buffers = new HashMap<>();
+  ConcurrentHashMap<String, Future> activeFlushFutures = new ConcurrentHashMap<>();
+
+  ScheduledExecutorService threadPool;
+  private static int shardCount;
+  private static int maxBufferSize;
+  private final Kryo kryo = new Kryo();
+  private static final String BUFFER_SHARD_PREFIX = "polaris-event-buffer-shard-";
+
+  @Inject
+  public FileBufferPolarisPersistenceEventListener(
+      MetaStoreManagerFactory metaStoreManagerFactory,
+      PolarisConfigurationStore polarisConfigurationStore,
+      Clock clock) {
+    this.metaStoreManagerFactory = metaStoreManagerFactory;
+    this.polarisConfigurationStore = polarisConfigurationStore;
+    this.clock = clock;
+    shardCount =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_NUM_SHARDS);
+    maxBufferSize =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_MAX_SIZE);
+    int timeToFlush =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_TIME_TO_FLUSH_IN_MS);
+    int numThreads =
+        polarisConfigurationStore.getConfiguration(
+            null, FeatureConfiguration.EVENT_BUFFER_NUM_THREADS);
+    threadPool = Executors.newScheduledThreadPool(numThreads);
+    kryo.register(PolarisEvent.class);
+    kryo.register(PolarisEvent.ResourceType.class);
+
+    // Start BufferListingTask
+    Function<FileBufferListingTask.TaskSubmissionInput, Future> taskSubmissionFunction =
+        input -> threadPool.schedule(input.task(), input.delayInMs(), TimeUnit.MILLISECONDS);
```

Comment:
There is no way here to cancel the `ScheduledFuture`. Are we OK with that?

Reply:
The design is not good then: you should always plan for the worse. Tasks _will_ stop by throwing unexpected runtime exceptions.